### DISCUSSION TITLE
Query API seems to just return lists of empty objects

### DISCUSSION THREAD

**Discussion by JAForbes** [open-policy-agent/community#442] [OPA and Rego] [answered]
Hey again!

I am positive I'm doing something wrong here, so apologies in advance!  But no matter what I try I can't seem to get any results from the query API beyond lists of empty objects.

I'm on 0.51.0 currently:

Here's a simple example:

```
POST http://localhost:8181/v1/query
Content-Type: application/json
{
  "input": { "example": "hello" },
  "query": "input.example"
}
```

I expect a response of `{ result: ["hello"] }` but I get back: `{ result: [{}] }`

I get similar results when I query data, if there's say 7 items in a JSON list, I get back 7 empty objects.

Thanks for the help :)

**Comment by srenatus** [comment, marked as answer]
Hello!

Your query is successful, but that's it. If you want to see the result of evaluation `input.example`, you'll have to bind it:

```interactive
$ curl http://127.0.0.1:8181/v1/query -d '{"input": {"example": "hello"}, "query": "x := input.example"}'
{"result":[{"x":"hello"}]}
$
```

**Reply by anderseknert** [reply]
This is a common source of confusion. Any reason we can't make this work like `opa eval` and just provide the result even when no binding is provided?

**Comment by JAForbes** [comment]
Thanks this worked! It did confuse me for a few hours so might be good to not require a binding in a future update.

Another small feature request would be to be able to specify a path to return on the resulting expressions/bindings so all intermediate bindings aren't sent back to the client.

E.g. if you had 3 intermediate variables in a query `x`, `y` and `z`, but after querying you only want `z.sub.property` it'd be great to be able to specify a path `http://localhost:8181/v1/query/z.sub.property`

**Reply by srenatus** [reply]
Interesting perspective. I feel like you've got a unique approach to OPA's HTTP API 😄 Most people push policies and query their rules, I think. You can do anything you want in a rule body, also return only `z.sub.property`, like
```rego
package example

query := z.sub.property {
  x := input.x
  y := input.y
  z := x[y] # you know, something complicated
}
```
and then you'd query `POST /v1/data/example/query` with payload `{"input": {"a": ..., "b": ... }}`

**Reply by JAForbes** [reply]
Hmm I can see the need to push policies into OPA server via the API.  But I still think it's pretty compelling to have a dynamic query API that doesn't necessarily send back each intermediate variable.
